packages/reputation-miner/ReputationMiner.js

```diff
@@ -758,12 +761,131 @@ class ReputationMiner {
+        }
       }
       if (applyLogs === false && localHash === hash) {
         applyLogs = true;
       }
     }
   }
+
+  async saveCurrentState() {
+    const db = await sqlite.open(this.dbPath, { Promise });
+
+    const currentRootHash = await this.getRootHash();
+    let res = await db.all(`SELECT rowid, root_hash FROM reputation_states WHERE root_hash='${currentRootHash}' AND n_nodes='${this.nReputations}'`);
+
+    let rootHashDBId;
+    if (res.length === 0) {
+      res = await db.run(`INSERT INTO reputation_states (root_hash, n_nodes) VALUES ('${currentRootHash}', ${this.nReputations})`);
+      rootHashDBId = res.lastID;
+    } else {
+      rootHashDBId = res[0].rowid;
+    }
+
+    for (let i = 0; i < Object.keys(this.reputations).length; i += 1) {
+      let colonyDBId;
+      let userDBId;
+      let skillDBId;
+      const key = Object.keys(this.reputations)[i];
+      const value = this.reputations[key];
+      let [colonyAddress, skillId, userAddress] = await ReputationMiner.breakKeyInToElements(key); // eslint-disable-line no-await-in-loop
+      colonyAddress = `0x${colonyAddress}`;
+      skillId = parseInt(skillId, 16);
+      userAddress = `0x${userAddress}`;
+
+      res = await db.all(`SELECT rowid FROM colonies WHERE address='${colonyAddress}'`); // eslint-disable-line no-await-in-loop
+      if (res.length === 0) {
+        res = await db.run(`INSERT INTO colonies (address) VALUES ('${colonyAddress}')`); // eslint-disable-line no-await-in-loop
+        colonyDBId = res.lastID;
+      } else {
+        colonyDBId = res[0].rowid;
+      }
+
+      res = await db.all(`SELECT rowid FROM users WHERE address='${userAddress}'`); // eslint-disable-line no-await-in-loop
+      if (res.length === 0) {
+        res = await db.run(`INSERT INTO users (address) VALUES ('${userAddress}')`); // eslint-disable-line no-await-in-loop
+        userDBId = res.lastID;
+      } else {
+        userDBId = res[0].rowid;
+      }
+
+      res = await db.all(`SELECT skill_id FROM skills WHERE skill_id='${skillId}'`); // eslint-disable-line no-await-in-loop
+      if (res.length === 0) {
+        res = await db.run(`INSERT INTO skills (skill_id) VALUES ('${skillId}')`); // eslint-disable-line no-await-in-loop
+        skillDBId = res.lastID;
+      } else {
+        skillDBId = res[0].skill_id;
+      }
+
+      // eslint-disable-next-line no-await-in-loop
+      res = await db.get(
+        `SELECT COUNT ( * ) AS "n"
+        FROM reputations
+        WHERE root_hash=${rootHashDBId}
+        AND colony_address=${colonyDBId}
+        AND skill_id=${skillDBId}
+        AND user_address=${userDBId}`
+      );
+
+      if (res.n === 0) {
+        // eslint-disable-next-line no-await-in-loop
+        await db.run(
+          `INSERT INTO reputations (root_hash, colony_address, skill_id, user_address, value)
+          VALUES (${rootHashDBId}, ${colonyDBId}, ${skillDBId}, ${userDBId}, '${value}')`
```

Comment:
We should really bundle up all `INSERT`s into a single transaction to ensure atomicity and data consistency. Documentation is missing on how to achieve that with `sqlite` but there is a test for it in https://github.com/mapbox/node-sqlite3/blob/113be5a8a6ec0e4f7c96045ab69119ea4e9a1427/benchmark/insert.js#L14

Reply:
What race condition are you trying to guard against here? We're not doing any queries simultaneously.

Reply:
It wasn't a race condition but if any of the `INSERT`s fails in `saveCurrentState` fails